Project.toml

```diff
@@ -4,6 +4,7 @@ authors = ["JuliaStats"]
 version = "0.25.29"
 
 [deps]
+ArraysOfArrays = "65a8f2f4-9b39-5baf-92e2-a9cc46fdf018"
```

Comment:
I don't think the implementation has to or should use `ArraysOfArrays` explicitly - I guess you can just work with generic `AbstractVector{<:AbstractVector}`, and then one can e.g. provide such Array wrappers from ArraysOfArrays if desired.

Reply:
Ok. But then there are some significant adjustments I have to make, cause I've relied on ArraysOfArrrays to do stuff like convert matrices to vector of vectors. What do you suggest in order to dispatch them?

Reply:
IMO we should remove the matrix dispatch (even regardless of ArraysOfArrays) as data can be organized in rows or columns - but this is not clear if just a matrix is provided. I think users should just use `eachrow`/`eachcol` (in a future Julia version or with Compat: https://github.com/JuliaLang/julia/pull/32310), ArraysOfArrays or ColVecs/RowVecs in KernelFunctions (I wanted to wait with extracting them a bit since I was hoping for the Julia PR).

Reply:
Ok, I'll drop the matrix dispatch.

Reply:
I just realized that there is an issue with not using `ArraysOfArrays`. I've heavily used it for computing `mean`, `var` and `cov`. I doesn't seem to work with a regular vector of vectors. Do you have a suggestion for it?

Reply:
I just realized that computations of weighted `mean`, `var` and `cov` depend on `ArraysOfArrays`. Do you know a workaround?

Reply:
They all work for vectors of vectors (of course, `ArraysOfArrays`, `ColVecs` or `EachCol` may implement more optimized versions of `mean` etc., but the default fallback works always):
```julia
julia> using Statistics

julia> x = [rand(5) for _ in 1:5];

julia> mean(x)
5-element Vector{Float64}:
 0.2609279080845831
 0.5184741034528969
 0.523898105188404
 0.6408166253497452
 0.33381236831836236

julia> var(x)
5-element Vector{Float64}:
 0.019751971071567172
 0.07655721742732527
 0.06221826760951234
 0.022761714823869204
 0.10639711437483416

julia> cov(x)
5×5 Matrix{Float64}:
  0.019752    -0.0182541  -0.00331084   0.0142255  -0.0231928
 -0.0182541    0.0765572  -0.0500437   -0.0402498   0.0677088
 -0.00331084  -0.0500437   0.0622183    0.0214676  -0.0425375
  0.0142255   -0.0402498   0.0214676    0.0227617  -0.0405306
 -0.0231928    0.0677088  -0.0425375   -0.0405306   0.106397
```

Reply:
Not the weighted version. I ended up writing a for-loop. Don't know if it's the most efficient way.

Reply:
Ah I didn't realize that you need the weighted ones. They are defined in StatsBase, there are some issues about exactly the vector of vectors case: https://github.com/JuliaStats/StatsBase.jl/issues/534 and https://github.com/JuliaStats/StatsBase.jl/issues/518

Reply:
Here is what I came up with:
```julia
function cov(d::MvDiscreteNonParametric)
    x = hcat(support(d)...)
    p = probs(d)
    return cov(x, Weights(p, one(eltype(p))), 2,corrected = false)
end
```

Reply:
@devmotion , are you fine with this implementation?